File: Lab3/main.c
//Assignment 3: Unsupervised Learning
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define ROWS 178
#define COLS 13

int file_reader(int classes[ROWS], double data[ROWS][COLS]) {
    FILE *fp;
    int i, j;

    fp = fopen("WINE.txt", "r");
    if (fp == NULL) {
        perror("Kunde inte öppna filen");
        return 1;
    }

    for (i = 0; i < ROWS; i++) {
        // Läs först klasskolumnen
        if (fscanf(fp, "%d", &classes[i]) != 1) {
            fprintf(stderr, "Fel vid läsning av klass på rad %d\n", i+1);
            return 1;
        }
        // Läs sedan features
        for (j = 0; j < COLS; j++) {
            if (fscanf(fp, "%lf", &data[i][j]) != 1) {
                fprintf(stderr, "Fel vid läsning av feature på rad %d kolumn %d\n", i+1, j+1);
                return 1;
            }
        }
    }

    fclose(fp);
    return 0;
}

int normilize_data(double data[ROWS][COLS]) {
    int j = 0;
    int i = 0;
    double sum = 0.0;
    double mean = 0.0;
    
    for(j = 0; j < COLS; j++){
        sum = 0.0;
        mean = 0.0;
         
        for (i = 0; i < ROWS; i++) {
            sum += data[i][j];
        }
        mean = sum / ROWS;
        for (i = 0; i < ROWS; i++) {
            data[i][j] -= mean;
        }
    }

    return 0;
}

compute_covariance(double data[ROWS][COLS], double cov[COLS][COLS]) {
    int i = 0;
    int j = 0;
    int k = 0;

    for (i = 0; i < COLS; i++) {
        for (j = 0; j < COLS; j++) {
            cov[i][j] = 0.0;
            
            for (k = 0; k < ROWS; k++) {
                cov[i][j] += data[k][i] * data[k][j];
            }
            cov[i][j] /= (ROWS - 1); 
        }
    }

    return 0;
}


int main() {
    

    // En kolumn för klass (int) och en matris för features (double)
    int classes[ROWS];
    double data[ROWS][COLS]; 
    double cov[COLS][COLS];

    file_reader(classes, data);
    normilize_data(data);
    
    ///*
    // Testutskrift: första raden
    printf("Klass: %d\n", classes[0]);
    printf("Features: ");
    for (int j = 0; j < COLS - 1; j++) {
        printf("%lf ", data[0][j]);
    }
    printf("\n");

    return 0;
    //*/
}

File: Lab3/main.py
# Assignment 3: Unsupervised Learning
import numpy as np
import os

ROWS = 178
COLS = 13

os.chdir(os.path.dirname(__file__))

def file_reader():
    classes = np.zeros(ROWS, dtype=int)
    data = np.zeros((ROWS, COLS), dtype=float)

    with open("WINE.txt", "r") as fp:
        for i, line in enumerate(fp):
            values = line.strip().split()
            if len(values) != COLS + 1:
                print(f"Fel vid läsning av rad {i+1}")
                continue

            classes[i] = int(values[0])
            data[i] = np.array(values[1:], dtype=float)

    return classes, data


def normalize_data(data):
    # Compute mean and standard deviation for each column
    mean = np.mean(data, axis=0)
    std = np.std(data, axis=0)
    
    # Avoid division by zero in case std is 0
    std[std == 0] = 1.0
    
    # Standardize the data
    normalized_data = (data - mean) / std
    return normalized_data



def compute_covariance(data):
    cov = np.zeros((COLS, COLS))
    for i in range(COLS):
        for j in range(COLS):
            cov[i, j] = np.sum(data[:, i] * data[:, j]) / (ROWS - 1)
    return cov


def compute_pca(cov, data):
    eigenvalues, eigenvectors = np.linalg.eig(cov)

    idx = np.argsort(eigenvalues)[::-1] 
    eigenvalues = eigenvalues[idx] 
    eigenvectors = eigenvectors[:, idx]

    new_features = np.dot(data, eigenvectors)

    return new_features, eigenvalues, eigenvectors

def kmeans(data, k=3, max_iters=100, tol=1e-4):
    np.random.seed(42)  
    centroids = data[np.random.choice(data.shape[0], k, replace=False)]

    for iteration in range(max_iters):
        # Step 1: Assign each point to the nearest centroid
        distances = np.linalg.norm(data[:, np.newaxis] - centroids, axis=2)  # Shape: (n_samples, k)
        labels = np.argmin(distances, axis=1)
        
        # Step 2: Update centroids as the mean of assigned points
        new_centroids = np.array([data[labels == i].mean(axis=0) for i in range(k)])
        
        # Step 3: Check for convergence
        if np.all(np.linalg.norm(new_centroids - centroids, axis=1) < tol):
            break
        
        centroids = new_centroids
    
    return labels, centroids




def main():
    classes, data = file_reader()
    data = normalize_data(data)
    cov = compute_covariance(data)
    new_features, eigenvalues, eigenvectors = compute_pca(cov, data)

    k = 3

    
    for n_components in [2, 5, 8]:
        reduced_data = new_features[:, :n_components]

        labels, centroids = kmeans(reduced_data, k=k)

        
        cluster_sizes = np.bincount(labels)

        
        print(f"\nK-means with top {n_components} principal components:")
        for i in range(k):
            print(f"  Cluster {i+1}: {cluster_sizes[i]} points")
        print(f"  Centroid shape: {centroids.shape}")  


if __name__ == "__main__":
    main()
